In [34]:
# Importações

import pandas as pd


from src.config import DADOS_ORIGINAIS, DADOS_TRATADOS

In [35]:
# Obtendo os dados como DataFrame
df = pd.read_csv(DADOS_ORIGINAIS)
df.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  2012-09-04       58       635  ...                  7             0   
1  2014-03-08       38        11  ...                  5             0   
2  2013-08-21       26       426  ...                  4             0   
3  2014-02-10       26        11  ...                  6             0   
4  2014-01-19       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

In [36]:
# Resumo da base
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [37]:
# Resumo estatístico das colunas numéricas
with pd.option_context(
    "display.max_columns", 26, "display.float_format", "{:.2f}".format
):
    display(df.describe())

ID  Year_Birth    Income  Kidhome  Teenhome  Recency  MntWines  \
count  2240.00     2240.00   2216.00  2240.00   2240.00  2240.00   2240.00   
mean   5592.16     1968.81  52247.25     0.44      0.51    49.11    303.94   
std    3246.66       11.98  25173.08     0.54      0.54    28.96    336.60   
min       0.00     1893.00   1730.00     0.00      0.00     0.00      0.00   
25%    2828.25     1959.00  35303.00     0.00      0.00    24.00     23.75   
50%    5458.50     1970.00  51381.50     0.00      0.00    49.00    173.50   
75%    8427.75     1977.00  68522.00     1.00      1.00    74.00    504.25   
max   11191.00     1996.00 666666.00     2.00      2.00    99.00   1493.00   

       MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
count    2240.00          2240.00          2240.00           2240.00   
mean       26.30           166.95            37.53             27.06   
std        39.77           225.72            54.63             41.28   
min         0.00             0.00             0.00              0.00   
25%         1.00            16.00             3.00              1.00   
50%         8.00            67.00            12.00              8.00   
75%        33.00           232.00            50.00             33.00   
max       199.00          1725.00           259.00            263.00   

       MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
count       2240.00            2240.00          2240.00              2240.00   
mean          44.02               2.33             4.08                 2.66   
std           52.17               1.93             2.78                 2.92   
min            0.00               0.00             0.00                 0.00   
25%            9.00               1.00             2.00                 0.00   
50%           24.00               2.00             4.00                 2.00   
75%           56.00               3.00             6.00                 4.00   
max          362.00              15.00            27.00                28.00   

       NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
count            2240.00            2240.00       2240.00       2240.00   
mean                5.79               5.32          0.07          0.07   
std                 3.25               2.43          0.26          0.26   
min                 0.00               0.00          0.00          0.00   
25%                 3.00               3.00          0.00          0.00   
50%                 5.00               6.00          0.00          0.00   
75%                 8.00               7.00          0.00          0.00   
max                13.00              20.00          1.00          1.00   

       AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  \
count       2240.00       2240.00       2240.00   2240.00        2240.00   
mean           0.07          0.06          0.01      0.01           3.00   
std            0.26          0.25          0.11      0.10           0.00   
min            0.00          0.00          0.00      0.00           3.00   
25%            0.00          0.00          0.00      0.00           3.00   
50%            0.00          0.00          0.00      0.00           3.00   
75%            0.00          0.00          0.00      0.00           3.00   
max            1.00          1.00          1.00      1.00           3.00   

       Z_Revenue  Response  
count    2240.00   2240.00  
mean       11.00      0.15  
std         0.00      0.36  
min        11.00      0.00  
25%        11.00      0.00  
50%        11.00      0.00  
75%        11.00      0.00  
max        11.00      1.00

### Aspectos interessantes

`Year_birth` com valor mínimo 1893, por ser um ano com diferença grande em relação ao da coleta da base. Possíveis hipóteses:

- Um parente usando a "conta" de um familiar provavelmente falecido
- Valor erroneamente imputado

`Income` com valor máximo 666666, por ser um valor muito específico e altamente discrepante em relação aos quartis. Isso será explorado nas análises de uma e duas variáveis.

`Z_CostContant` e `Z_revenue` com quartis e mínimo e máximo iguais, que não agregarão valor a análise 


In [38]:
# Resumo estatístico das colunas categóricas
df.describe(exclude="number")

Education Marital_Status Dt_Customer
count         2240           2240        2240
unique           5              8         663
top     Graduation        Married  2012-08-31
freq          1127            864          12

### Aspectos interessantes

- Quase metade das ocorrências do atributo `Marital_Status` ser 'Graduation'

### Atividades de limpeza e transformação

- Analisar o valor 1893 e outros possíveis valores estranhos em `Year_Birth`
- Verificar a padronização das datas de `Dt_Customer`
- Converter `Dt_Customer` para datetime
- Verificar se as datas de `Dt_Customer` são válidas
- Verificar se há incosistência de datas entre `Year_Birth` e `Dt_Customer`, ou seja, se o ano de ingressão do cliente é menor do que o ano de nascimento do cliente
- Analisar inconsistência de rótulos em `Response` de registros duplicados

In [39]:
# Verificando a quantidade de valores únicos das colunas
df.nunique()

ID                     2240
Year_Birth               59
Education                 5
Marital_Status            8
Income                 1974
Kidhome                   3
Teenhome                  3
Dt_Customer             663
Recency                 100
MntWines                776
MntFruits               158
MntMeatProducts         558
MntFishProducts         182
MntSweetProducts        177
MntGoldProds            213
NumDealsPurchases        15
NumWebPurchases          15
NumCatalogPurchases      14
NumStorePurchases        14
NumWebVisitsMonth        16
AcceptedCmp3              2
AcceptedCmp4              2
AcceptedCmp5              2
AcceptedCmp1              2
AcceptedCmp2              2
Complain                  2
Z_CostContact             1
Z_Revenue                 1
Response                  2
dtype: int64

A cardinalidade do atributo `ID` igual ao número de registros sugere que, assim como `Z_CostContant` e `Z_revenue`, ela não agregará valor a análise. Portanto, todas elas podem ser removidas.

In [40]:
# Removendo as colunas ID, 'Z_CostContact' e 'Z_Revenue'
df = df.drop(['ID', 'Z_CostContact', 'Z_Revenue'], axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Year_Birth           2240 non-null   int64  
 1   Education            2240 non-null   object 
 2   Marital_Status       2240 non-null   object 
 3   Income               2216 non-null   float64
 4   Kidhome              2240 non-null   int64  
 5   Teenhome             2240 non-null   int64  
 6   Dt_Customer          2240 non-null   object 
 7   Recency              2240 non-null   int64  
 8   MntWines             2240 non-null   int64  
 9   MntFruits            2240 non-null   int64  
 10  MntMeatProducts      2240 non-null   int64  
 11  MntFishProducts      2240 non-null   int64  
 12  MntSweetProducts     2240 non-null   int64  
 13  MntGoldProds         2240 non-null   int64  
 14  NumDealsPurchases    2240 non-null   int64  
 15  NumWebPurchases      2240 non-null   i

In [41]:
# Verificando os valores nulos da base
with pd.option_context("display.max_columns", 29):

    display(df[df.isnull().any(axis=1)])

Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
10          1983  Graduation        Married     NaN        1         0   
27          1986  Graduation         Single     NaN        1         0   
43          1959         PhD         Single     NaN        0         0   
48          1951  Graduation         Single     NaN        2         1   
58          1982  Graduation         Single     NaN        1         0   
71          1973    2n Cycle        Married     NaN        1         0   
90          1957         PhD        Married     NaN        2         1   
91          1957  Graduation         Single     NaN        1         1   
92          1973      Master       Together     NaN        0         0   
128         1961         PhD        Married     NaN        0         1   
133         1963  Graduation        Married     NaN        0         1   
312         1989  Graduation        Married     NaN        0         0   
319         1970  Graduation         Single     NaN        1         2   
1379        1970      Master       Together     NaN        0         1   
1382        1958  Graduation       Together     NaN        1         1   
1383        1964    2n Cycle         Single     NaN        1         1   
1386        1972         PhD       Together     NaN        1         0   
2059        1969      Master       Together     NaN        1         1   
2061        1981         PhD         Single     NaN        1         0   
2078        1971  Graduation        Married     NaN        1         1   
2079        1954      Master       Together     NaN        0         1   
2081        1955  Graduation         Single     NaN        0         1   
2084        1943      Master          Widow     NaN        0         0   
2228        1978    2n Cycle       Together     NaN        0         0   

     Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
10    2013-11-15       11         5          5                6   
27    2013-02-20       19         5          1                3   
43    2013-11-05       80        81         11               50   
48    2014-01-01       96        48          5               48   
58    2013-06-17       57        11          3               22   
71    2012-09-14       25        25          3               43   
90    2012-11-19        4       230         42              192   
91    2014-05-27       45         7          0                8   
92    2013-11-23       87       445         37              359   
128   2013-07-11       23       352          0               27   
133   2013-08-11       96       231         65              196   
312   2013-06-03       69       861        138              461   
319   2013-08-23       67       738         20              172   
1379  2013-04-01       39       187          5               65   
1382  2012-09-03       87        19          4               12   
1383  2014-01-12       49         5          1                9   
1386  2014-03-02       17        25          1               13   
2059  2013-05-18       52       375         42               48   
2061  2013-05-31       82        23          0               15   
2078  2013-03-03       82        71          1               16   
2079  2013-06-23       83       161          0               22   
2081  2013-10-18       95       264          0               21   
2084  2013-10-30       75       532        126              490   
2228  2012-08-12       53        32          2             1607   

      MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
10                  0                 2             1                  1   
27                  3               263           362                  0   
43                  3                 2            39                  1   
48                  6                10             7                  3   
58                  2                 2             6                  2   
71                 17  

Todas os registros que possuem valores nulos em `Income` se referem a clientes que não aceitaram as ofertas das campanhas 2 e 3 (`AcceptedCmp2` e `AcceptedCmp3` iguais a 0). 

In [42]:
# Média dos clientes que não aceitaram as ofertas das campanhas 2 e 3
df[(df["Income"].notnull()) & (df["AcceptedCmp2"] == 0) & (df["AcceptedCmp2"] == 0)]["Income"].median()

np.float64(51117.5)

Usaremos a mediana de `Income` desses registros para imputar os valores ausentes de `Income`, para que o valor não seja sensível a outliers

In [43]:
# Imputando a mediana de income para registros com essas características

median_income = df[(df["Income"].notnull()) & (df["AcceptedCmp2"] == 0) & (df["AcceptedCmp2"] == 0)]["Income"].median()
df.loc[df["Income"].isnull(), "Income"] = median_income

In [44]:
# Verificando se ainda há valores nulos na base
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Complain, Response]
Index: []

[0 rows x 26 columns]

In [45]:
# Verificando o registro com Year_Birth igual a 1893
with pd.option_context("display.max_columns", 29):

    display(df[df["Year_Birth"] == 1893])

Year_Birth Education Marital_Status   Income  Kidhome  Teenhome  \
239        1893  2n Cycle         Single  60182.0        0         1   

    Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
239  2014-05-17       23         8          0                5   

     MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
239                7                 0             2                  1   

     NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
239                1                    0                  2   

     NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
239                  4             0             0             0   

     AcceptedCmp1  AcceptedCmp2  Complain  Response  
239             0             0         0         0

In [46]:
# Verificando os menores valores de Year_Birth e suas ocorrências
df["Year_Birth"].value_counts().sort_index().head()

Year_Birth
1893    1
1899    1
1900    1
1940    1
1941    1
Name: count, dtype: int64

In [47]:
# Verificando o registros com Year_Birth menores que 1900
with pd.option_context("display.max_columns", 29):

    display(df[df["Year_Birth"] <= 1900])

Year_Birth Education Marital_Status   Income  Kidhome  Teenhome  \
192        1900  2n Cycle       Divorced  36640.0        1         0   
239        1893  2n Cycle         Single  60182.0        0         1   
339        1899       PhD       Together  83532.0        0         0   

    Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
192  2013-09-26       99        15          6                8   
239  2014-05-17       23         8          0                5   
339  2013-09-26       36       755        144              562   

     MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
192                7                 4            25                  1   
239                7                 0             2                  1   
339              104                64           224                  1   

     NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
192                2                    1                  2   
239                1                    0                  2   
339                4                    6                  4   

     NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
192                  5             0             0             0   
239                  4             0             0             0   
339                  1             0             0             1   

     AcceptedCmp1  AcceptedCmp2  Complain  Response  
192             0             0         1         0  
239             0             0         0         0  
339             0             0         0         0

Existem três registros de clientes com `Year_Birth` estranhos. Dentre as hipóteses listadas, não há uma forma de comprovar nenhuma das duas. Porém, como são poucos registros, opto por removê-los da base

In [48]:
# Removendo os registros com Year_Birth menores que 1900
df = df[df["Year_Birth"] > 1900]
df[df["Year_Birth"] <= 1900]

Empty DataFrame
Columns: [Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Complain, Response]
Index: []

[0 rows x 26 columns]

In [49]:
df["Dt_Customer"]

0       2012-09-04
1       2014-03-08
2       2013-08-21
3       2014-02-10
4       2014-01-19
           ...    
2235    2013-06-13
2236    2014-06-10
2237    2014-01-25
2238    2014-01-24
2239    2012-10-15
Name: Dt_Customer, Length: 2237, dtype: object

In [50]:
# Verificando a padronização das data em Dt_Customer
df["Dt_Customer"].str.contains("\d\d\d\d-\d\d-\d\d", regex=True).sum()

np.int64(2237)

In [51]:
# Convertendo a coluna Dt_Customer para datetime
df["Dt_Customer"] = pd.to_datetime(df["Dt_Customer"], format="%Y-%m-%d")
df["Dt_Customer"].dtype

dtype('<M8[ns]')

In [52]:
# Verificando as datas mais antigas
df["Dt_Customer"].nsmallest(5)

1255   2012-07-30
1363   2012-07-31
724    2012-08-01
976    2012-08-01
2029   2012-08-01
Name: Dt_Customer, dtype: datetime64[ns]

In [53]:
# Verificando as datas mais novas
df["Dt_Customer"].nlargest(5)

200    2014-06-29
989    2014-06-29
45     2014-06-28
796    2014-06-28
1812   2014-06-28
Name: Dt_Customer, dtype: datetime64[ns]

In [54]:
# Verificando se não há inconsistências entre Year_Birth e o ano de Dt_Customer
df[df["Year_Birth"] >= df["Dt_Customer"].dt.year]

Empty DataFrame
Columns: [Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Complain, Response]
Index: []

[0 rows x 26 columns]

Não parece haver problemas de padronização de datas, nem datas errôneas em `Dt_Customer`, e também nenhuma inconsistência com `Year_Birth`

No entanto, o algoritmo de classificação pode ter problemas com relação ao tipo do atributo `Dt_Customer` ser datetime. Logo, é interessante gerar atributos derivados desse atributo.

Vou gerar um atributo numérico referente aos dias desde a inscrição do cliente (`DaysSinceEnrolled`) e um atributo categórico referente aos anos desde a inscrição do cliente (`YearsSinceEnrolled`), ambos com referência a data máxima de `Dt_Customer`.

In [55]:
# Gerando o atributo DaysSinceEnrolled a partir de Dt_Customer
df["DaysSinceEnrolled"] = (df["Dt_Customer"].max() - df["Dt_Customer"]).dt.days
df["DaysSinceEnrolled"].head()

0    663
1    113
2    312
3    139
4    161
Name: DaysSinceEnrolled, dtype: int64

In [56]:
# Gerando o atributo YearsSinceEnrolled a partir de Dt_Customer
df["YearsSinceEnrolled"] = df["DaysSinceEnrolled"] // 365
df["YearsSinceEnrolled"].head()

0    1
1    0
2    0
3    0
4    0
Name: YearsSinceEnrolled, dtype: int64

In [57]:
# Verificando a cardinalidade de YearsSinceEnrolled
df["YearsSinceEnrolled"].nunique()

2

Também vou gerar um atributo derivado de `Year_Birth`, correspondente a idade do cliente em relação ao ano da data máxima de inscrição. Nesse caso, o algoritmo não vai ter tantos ganhos, mas pode ser mais interpretável para uma pessoa.

In [58]:
# Gerando o atributo Age a partir de Year_Birth
df["Age"] = df["Dt_Customer"].max().year - df["Year_Birth"]
df["Age"].head()

0    57
1    60
2    49
3    30
4    33
Name: Age, dtype: int64

In [59]:
# Verificando se há registros duplicados
df.duplicated().sum()

np.int64(182)

In [60]:
# Verificando a quantidade de registros da classe minoritária em Response
df[(df.duplicated()) & (df["Response"] == 1)].shape[0]

21

Registros duplicados não é necessariamente um problema para o algoritmo de classificação. Possivelmente, eles podem mascarar as métricas de avaliação do modelo, mas reduzir ainda mais a quantidade de registros da classe minoritária é pior ainda.

No entanto, registros duplicados com rótulos no atributo alvo inconsistentes será um grave problema, pois o algoritmo pode capturar os padrões de ambos os registros no treino e classificar erroneamente no teste. Esses registros duplicados, em especifico, serão removidos.

In [61]:
# Verificando os registros duplicados que apresentam rótulos diferentes (inconsistência)
df.groupby(
    df.columns.drop("Response").tolist()
    ).filter(
        lambda x: x["Response"].nunique() >= 2
        )

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
8           1974         PhD       Together  30351.0        1         0   
84          1987  Graduation       Divorced  81361.0        0         0   
141         1976      Master       Together  65104.0        0         1   
195         1977  Graduation        Married  31353.0        1         1   
256         1961         PhD       Together  63342.0        0         1   
374         1972         PhD         Single  38578.0        1         1   
375         1975      Master         Single  46098.0        1         1   
408         1974         PhD       Together  30351.0        1         0   
486         1976      Master       Together  65104.0        0         1   
548         1968  Graduation        Married  19514.0        1         1   
690         1960         PhD       Divorced  50611.0        0         1   
852         1957    2n Cycle        Married  82347.0        0         0   
912         1963         PhD          Widow  52278.0        0         1   
968         1974         PhD        Married  34596.0        1         0   
1008        1968  Graduation        Married  19514.0        1         1   
1048        1984  Graduation       Together  35196.0        1         0   
1060        1949      Master        Married  62845.0        1         1   
1063        1973         PhD       Divorced  35860.0        1         1   
1093        1972         PhD         Single  38578.0        1         1   
1179        1991      Master         Single  90638.0        0         0   
1200        1977  Graduation        Married  54809.0        1         1   
1256        1955    2n Cycle        Married  62972.0        0         1   
1260        1973         PhD       Divorced  35860.0        1         1   
1342        1949      Master        Married  62845.0        1         1   
1373        1960         PhD       Divorced  50611.0        0         1   
1426        1974         PhD        Married  34596.0        1         0   
1511        1975      Master         Single  46098.0        1         1   
1533        1975      Master         Single  46098.0        1         1   
1572        1991      Master         Single  90638.0        0         0   
1584        1987  Graduation       Divorced  81361.0        0         0   
1585        1973         PhD       Divorced  35860.0        1         1   
1614        1977  Graduation        Married  31353.0        1         1   
1715        1984  Graduation       Together  35196.0        1         0   
1750        1961         PhD       Together  63342.0        0         1   
1834        1977  Graduation        Married  54809.0        1         1   
1838        1955    2n Cycle        Married  62972.0        0         1   
1914        1963         PhD          Widow  52278.0        0         1   
2177        1973         PhD           YOLO  48432.0        0         1   
2193        1957    2n Cycle        Married  82347.0        0         0   
2202        1973         PhD           YOLO  48432.0        0         1   

     Dt_Customer  Recency  MntWines  MntFruits  ...  AcceptedCmp3  \
8     2013-06-06       19        14          0  ...             0   
84    2014-02-25       18       163         23  ...             0   
141   2013-11-14        4       738          0  ...             0   
195   2013-12-29       24        10          0  ...             0   
256   2012-10-16       48       918         21  ...             0   
374   2013-06-22        2        38          4  ...             0   
375   2012-08-18       86        57          0  ...             0   
408   2013-06-06       19        14          0  ...             0   
486   2013-11-14        4       738          0  ...             0   
548   2014-01-26       47        14          3  ...             1   
690   2012-10-04       98       459          0  ...             0   
852   2012-11-06       38       556         54  ...             1   
912   2013-01-25       24       9

In [62]:
# Armazenando os indices registros duplicados com rótulos inconsistentes
inconsistences_indexes = df.groupby(
    df.columns.drop("Response").tolist()
    ).filter(
        lambda x: x["Response"].nunique() >= 2
        ).index
inconsistences_indexes

Index([   8,   84,  141,  195,  256,  374,  375,  408,  486,  548,  690,  852,
        912,  968, 1008, 1048, 1060, 1063, 1093, 1179, 1200, 1256, 1260, 1342,
       1373, 1426, 1511, 1533, 1572, 1584, 1585, 1614, 1715, 1750, 1834, 1838,
       1914, 2177, 2193, 2202],
      dtype='int64')

In [63]:
# Removendo esses registros duplicados inconsistentes
df = df[~(df.index.isin(inconsistences_indexes))].reset_index(drop=True)

In [64]:
# Verificando se ainda há registros duplicados inconsistentes
df.groupby(
    df.columns.drop("Response").tolist()
    ).filter(
        lambda x: x["Response"].nunique() >= 2
        )

Empty DataFrame
Columns: [Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Complain, Response, DaysSinceEnrolled, YearsSinceEnrolled, Age]
Index: []

[0 rows x 29 columns]

In [65]:
# Verificando a proporção do atributo alvo
df["Response"].value_counts(normalize=True)

Response
0    0.856623
1    0.143377
Name: proportion, dtype: float64

In [66]:
# Salvando as mudanças
df.to_parquet(DADOS_TRATADOS, index=False)